# FinInsights: LLM-Powered Financial Assistant (Demo)

This notebook demonstrates a simple LLM-powered assistant that can answer natural language queries about financial data such as portfolio allocations, sector performance, or company fundamentals.

We use the OpenAI API to power a conversational assistant that can:
- Interpret structured data (CSV-based financial or portfolio data)
- Generate textual summaries
- Provide sector insights from plain English queries

This is a simple prototype version designed for demonstration in a single notebook.

### Setup & Data Load

In [74]:
import os
from openai import OpenAI
from dotenv import load_dotenv
import requests
import httpx

# Load API key and proxy from .env
load_dotenv()
api_key = os.getenv("OPENAI_API_KEY")
http_proxy = os.getenv("HTTP_PROXY")
https_proxy = os.getenv("HTTPS_PROXY")

proxies = {
    "http": http_proxy,
    "https": https_proxy
}

# Load sample portfolio data
df = pd.read_csv("sample_portfolio_30.csv", encoding="utf-8", encoding_errors="ignore")
print(df.head())
print(df.shape)

    Symbol                                          Name  Allocation  \
0  6889106  Taiwan Semiconductor Manufacturing Co., Ltd.    0.269198   
1  BMMV2K8                          Tencent Holdings Ltd    0.120909   
2  BK6YZP5                 Alibaba Group Holding Limited    0.071246   
3  6771720                 Samsung Electronics Co., Ltd.    0.068551   
4  BK1N461                             HDFC Bank Limited    0.042680   

   Adjusted_Market_Value  Dividend_Yield  Price_to_Earnings  Debt_to_Equity  \
0                 957463            1.30              26.87           24.49   
1                 451376            0.82              21.09           36.78   
2                 253403            1.18              16.97           24.57   
3                 243820            2.71              11.27            4.94   
4                 130462            1.10              19.48          129.46   

  Country                  Sector                                   Industry  \
0  Taiwan  I

### Create Smaller Custom Dataframes

In [75]:
# Truncate Business_Description to 40 words
df["Short_Description"] = df["Business_Description"].apply(lambda x: " ".join(str(x).split()[:40]) + "...")

df_allocations = df[["Name", "Allocation", "Sector"]]
df_perf = df[["Name", "Allocation", "Country", "Sector", "YTD_Performance"]]
df_perf_fundamentals = df[["Name", "Allocation", "Country", "Sector", "YTD_Performance", "Adjusted_Market_Value", "Dividend_Yield", "Price_to_Earnings", "Debt_to_Equity"]]
df_business_desc = df[["Name", "Country", "Sector", "Short_Description"]]

### LLM Prompt Function

In [77]:
def ask_fininsights(data: pd.DataFrame, question: str, role, model="gpt-4"):
    """Send natural language question and tabular data to OpenAI via Chat API"""

    # Convert the entire DataFrame to markdown (table format) for inclusion in the prompt
    # Consider limiting rows (e.g., with data.head(10)) to avoid token limits
    data_sample = data.to_markdown(index=False)

    # Build the prompt to send to the model, including dataset and natural language question
    prompt = f"""
You are a financial analyst assistant. Given the following dataset:

{data_sample}

Answer this question clearly and concisely:
{question}
"""

    # Set headers for OpenAI API request, including authorization
    headers = {
        "Authorization": f"Bearer {api_key}",
        "Content-Type": "application/json"
    }

    # Construct the payload to send to OpenAI
    payload = {
        "model": model,
        "messages": [
            {"role": "system", "content": role},
            {"role": "user", "content": prompt}
        ],
        "temperature": 0.4,   # Lower temperature for more deterministic output
        "max_tokens": 500     # Limit output length to avoid excess token usage
    }

    try:
        # Make the POST request to OpenAI API with optional proxy and timeout
        response = requests.post(
            "https://api.openai.com/v1/chat/completions",
            headers=headers,
            json=payload,
            proxies=proxies,
            timeout=60
        )

        # If successful, return the LLM’s generated message content
        if response.status_code == 200:
            return response.json()["choices"][0]["message"]["content"]
        else:
            # Print the error message if API returns non-200 status
            print(f"OpenAI API Error: {response.status_code} - {response.text}")
            return None

    except requests.exceptions.RequestException as e:
        # Handle timeout, connection, or other request-related errors
        print(f"Request error: {e}")
        return None

### Simple Zero Shot Prompts

In [78]:
example_question_1 = "Which sectors have the highest allocations in the portfolio? Provide allocations."
example_question_2 = "What are top 3 performing companies (with their YTD Performance)?"
example_question_3 = "What are the 3 most similar companies to Alibaba based on business description?"

role = "You are a portfolio analyst specialising in emerging markets equities."

### Example 1: Sector Allocations
**Prompt:** "Which sectors have the highest allocations in the portfolio? Provide allocations."  

In [79]:
print("\n\nExample 1:")
print(ask_fininsights(df_allocations, example_question_1, role))



Example 1:
The sectors with the highest allocations in the portfolio are:

1. Information Technology: The total allocation is 0.269198 (Taiwan Semiconductor Manufacturing Co., Ltd.) + 0.0685511 (Samsung Electronics Co., Ltd.) + 0.0325908 (Xiaomi Corporation Class B) + 0.0302744 (SK hynix Inc.) + 0.0209634 (Hon Hai Precision Industry Co., Ltd.) + 0.0196372 (MediaTek Inc) + 0.0160924 (Infosys Limited) + 0.010404 (Delta Electronics, Inc.) = 0.4677111

2. Financials: The total allocation is 0.0426797 (HDFC Bank Limited) + 0.0277273 (China Construction Bank Corporation Class H) + 0.0250624 (ICICI Bank Limited) + 0.0140366 (Industrial and Commercial Bank of China Limited Class H) + 0.0139188 (Al Rajhi Bank) + 0.0129511 (Ping An Insurance (Group) Company of China, Ltd. Class H) + 0.0115572 (Bank of China Limited Class H) + 0.0110032 (Nu Holdings Ltd. Class A) = 0.1589363

3. Consumer Discretionary: The total allocation is 0.0712461 (Alibaba Group Holding Limited) + 0.0221492 (PDD Holdings I

### Example 2: Top Performing Companies  
**Prompt:** "What are top 3 performing companies (with their YTD Performance)?"

In [80]:
print("\n\nExample 2:")
print(ask_fininsights(df_perf, example_question_2, role))



Example 2:
The top 3 performing companies based on their Year-to-Date (YTD) performance are:

1. Delta Electronics, Inc. with a YTD performance of 59.5656%
2. SK hynix Inc. with a YTD performance of 58.7134%
3. Xiaomi Corporation Class B with a YTD performance of 54.8867%


### Example 3: Similar Businesses
**Prompt:** "What are the 3 most similar companies to Alibaba based on business description?"

In [81]:
print("\n\nExample 3:")
print(ask_fininsights(df_business_desc, example_question_3, role)) 



Example 3:
Based on the business descriptions provided, the three most similar companies to Alibaba Group Holding Limited would be:

1. Tencent Holdings Ltd: Similar to Alibaba, Tencent provides a variety of online and mobile services, including value-added services, online advertising services, and fintech and business services. Both companies have a broad range of operations in the technology sector.

2. Meituan Class B: Meituan, like Alibaba, operates a technology platform that connects consumers and merchants, providing a variety of services. Both companies are involved in the consumer discretionary sector and have a strong focus on leveraging technology to enhance commerce.

3. JD.com, Inc. Class A: JD.com is an e-commerce company that sells a wide range of products, similar to Alibaba. Both companies operate in the consumer discretionary sector and have a significant online retail presence.


# Prompt Engineering Examples

### Zero-shot
See above examples

### Role prompting:
See above examples with role = "You are a portfolio analyst specialising in emerging markets equities."
Improves tone, technical accuracy, and domain assumptions.

### Few-shot prompting:
Q: Which companies have the highest market value?  
A: 
1. TSMC - 957,463,     
2. Tencent – 451,376,  
4. Alibaba – 253,403,  

Q: Which companies have the lowest P/E ratios?  
A: 
1. Samsung Electronics – 11.27,
2. Delta Electronics – 13.45,  
3. Bank of China – 13.90,  

Q: What are the top 3 performing companies (YTD)?

### Format instruction
List the top 5 companies by allocation and include their ticker, country, and YTD performance. Format as a markdown table.

### Chain-of-thought reasoning
Which 3 companies have the strongest combination of low P/E, high dividend yield, and strong YTD performance? Show your reasoning.

### Scoring or Classification
For each company, classify whether its business is primarily domestic or international, based on its business description

In [83]:
example_question_4 = "List the top 5 companies by allocation and include their country, and YTD performance. Format as a markdown table."
example_question_5 = "Which 3 companies have the strongest combination of low P/E, high dividend yield, and strong YTD performance? Show your reasoning."
example_question_6 = "For each company, classify whether its business is primarily domestic or international, based on its business description"

print("\n\nFormat instruction example:")
print(ask_fininsights(df_perf, example_question_4, role))

print("\n\nChain-of-thought example:")
print(ask_fininsights(df_perf_fundamentals, example_question_5, role))

print("\n\nClassification example:")
print(ask_fininsights(df_business_desc.head(5), example_question_6, role))



Format instruction example:
Sure, here are the top 5 companies by allocation:

| Name                                                     |   Allocation | Country      |   YTD_Performance |
|:---------------------------------------------------------|-------------:|:-------------|------------------:|
| Taiwan Semiconductor Manufacturing Co., Ltd.             |   0.269198   | Taiwan       |         0.153872  |
| Tencent Holdings Ltd                                     |   0.120909   | China        |         0.359747  |
| Alibaba Group Holding Limited                            |   0.0712461  | China        |         0.425033  |
| Samsung Electronics Co., Ltd.                            |   0.0685511  | Korea        |         0.384884  |
| HDFC Bank Limited                                        |   0.0426797  | India        |         0.106574  |


Chain-of-thought example:
The three companies with the strongest combination of low Price to Earnings (P/E), high Dividend Yield, and strong

# Conclusion

This notebook explored using OpenAI's API to generate natural language insights from portfolio data. I tested the approach on a range of questions - including sector allocations, top performers, and company similarity based on business descriptions. I also experimented with different prompting styles—including zero-shot, role prompting, few-shot examples, formatting instructions, and reasoning-based queries—to evaluate how these affect the quality and reliability of the outputs.

The model was generally able to extract and summarise useful information, particularly for questions involving rankings or comparisons (e.g. top performers). It handled business descriptions quite well when asked about company similarities. However, the accuracy of numerical breakdowns (like allocation totals) was mixed, and explanations were often generic or shallow.

One key limitation is that the model only sees a subset of the data due to token limits. This means answers may miss context or appear biased towards the first few rows. It's also worth noting that the model isn’t accessing real-time news or external information - it's only working off the data I provided.

That said, it shows promise as a lightweight assistant for interpreting structured data. Future iterations could include pre-processed summaries (e.g. sector totals), embedding-based similarity rather than text matching, and integration with a wider data pipeline (e.g. newsflow, fundamentals, macro data).